File: FizzBuzz/TDD/JavaScript/src/FilterJob.js
const filtersProvider = require('./FiltersProvider').filtersProvider

filterFun = (input) => {
    if (input instanceof Array) {
	return input.map(it => tt(it))
    }
    return tt(input)
}

tt = (input) => {
    const result = applyFilters(input, filtersProvider.all(), '')
    if ('' === result.trim()) {
	return input.toString()
    }
    return result
}

applyFilters = (input, filters, result) => {
    throwIntegerLessThanOneException(input)
    if(0 === filters.size)
    {
	return result
    }
    const nextFilter = filters[Symbol.iterator]().next().value
    const partAnswer = nextFilter(input)
    filters.delete(nextFilter)
    return applyFilters(input, filters, result.concat(partAnswer))
}

throwIntegerLessThanOneException = (input) => {
    if (1 > input) {
	throw new IntegerLessThanOneException(input)
    }
}

module.exports = { filterJob: {  process: filterFun } }

File: FizzBuzz/TDD/erlang/src/FilterJob.erl
-module('FilterJob').

-export([filter/1]).

filter(Input) ->
    Result = applyFilters(Input),
    case length(Result) of
	0 ->
	    integer_to_list(Input);
	_ ->
	    Result
    end.

applyFilters(Input) ->
    Filters =  'FiltersProvider':filters(),
    applyFilters(Input, Filters, "").

applyFilters(_Input, [], Result) ->
	Result;
applyFilters(Input, [NextFilter|Filters], Result) ->
	applyFilters(Input, Filters, lists:concat([NextFilter(Input), Result])).
